feat(cli): support manual approvals in pipeline manifest.

tam0ri opened the pull request:
related #1179

When a production environment is added via `pipeline init`, copilot adds the requires_approval: true line in pipeline.yml.
When requires_approval: true line exists in pipeline.yml, manual approval action is added into the pipeline via `pipeline update`.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

tam0ri committed:
feat(cli): support manual approvals in pipeline manifest.

When a production environment is added via pipeline init, copilot adds the requires_approval: true line in pipeline.yml.
When requires_approval: true line exists in pipeline.yml, manual approval action is added into the pipeline via pipeline update.

tam0ri commented:
Hi reviewers. I'm not sure we need to update the version of pipeline manifest.

https://github.com/aws/copilot-cli/blob/a145f2271b8b294bbbd9e4faedbf979b44728034/internal/pkg/manifest/pipeline.go#L75-L78
https://github.com/aws/copilot-cli/blob/a145f2271b8b294bbbd9e4faedbf979b44728034/internal/pkg/manifest/pipeline.go#L119

If we need to increment the version, please tell me that.

efekarakus commented on internal/pkg/manifest/pipeline.go:
```diff
@@ -97,21 +98,26 @@ type Source struct {
 
 // CreatePipeline returns a pipeline manifest object.
-func CreatePipeline(pipelineName string, provider Provider, stageNames []string) (*PipelineManifest, error) {
+func CreatePipeline(pipelineName string, provider Provider, envs []*config.Environment) (*PipelineManifest, error) {
```
nit: Would you mind renaming this function to `NewPipelineManifest`?

tam0ri replied:
I addressed.

efekarakus commented on internal/pkg/cli/pipeline_init.go:
```diff
@@ -247,14 +247,36 @@ func (o *initPipelineOpts) createPipelineProvider() (manifest.Provider, error) {
+		environments = append(environments, env)
+	}
+
+	manifest, err := manifest.CreatePipeline(pipelineName, provider, environments)
```
Instead of passing `[]*config.Environment` as input to this function, what do you think of passing `[]manifest.PipelineStage`? so this method would transform `[]*config.Environment -> []manifest.PipelineStage`.

Since the function is in the `manifest` pkg, accepting `[]manifest.PipelineStage` will make it more flexible. The `PipelineStage` can be created from a `config.Environment` object or some other way.

tam0ri replied:
I addressed. However, now TestNewPipelineManifest just become meaningless because this test expects PipelineStage. I plan to change this test such as it expects PipelineManifest. If you have any opinion, please give me a feedback!

efekarakus replied:
hmm yeah I agree, the only behavior that might be interesting to test in `TestNewPipelineManifest` is checking if an error is returned if the list is empty and that's it

tam0ri replied:
Oh, does it mean we simply need to remove happy case and retain only error case? I just modified happy case with `PipelineManifest` as expected by imitating backend_svc_test.go. If I need to remove happy case, please point out!

efekarakus commented on internal/pkg/cli/pipeline_init.go:
```diff
@@ -247,14 +247,36 @@ func (o *initPipelineOpts) createPipelineProvider() (manifest.Provider, error) {
 	return manifest.NewProvider(config)
 }
 
+func (o *initPipelineOpts) getEnvFromCache(environmentName string) (*config.Environment, error) {
```
nit: Can we rename this method to `getEnvConfig`?

tam0ri replied:
I addressed.

efekarakus commented on internal/pkg/cli/pipeline_init.go:
```diff
@@ -247,14 +247,36 @@ func (o *initPipelineOpts) createPipelineProvider() (manifest.Provider, error) {
+	var environment *config.Environment
+	for _, env := range o.envs {
+		if env.Name == environmentName {
+			environment = env
```
nit: `break` after this statement so that we don't continue looping.

tam0ri replied:
I addressed.

efekarakus replied:
nit: oh my bad, instead of breaking we should just return instead :)

efekarakus replied:
and then outside of the for-loop we wouldn't need to check if it's equal to nil and return also an error immediately

tam0ri replied:
Totally agree! I addressed.

efekarakus reviewed (commented):
😍 Thank you so much! 

It looks good, mostly a comment about decoupling the `manifest` pkg from `config`.

efekarakus commented:
👋 @git823 

> If we need to increment the version, please tell me that.

We can keep it as is. We don't do anything with the version atm, so it should be fine :)

efekarakus commented on internal/pkg/cli/pipeline_update.go:
```diff
@@ -112,9 +112,9 @@ func (o *updatePipelineOpts) convertStages(manifestStages []manifest.PipelineSta
-				Prod:      env.Prod,
 			},
-			TestCommands: stage.TestCommands,
+			RequiresApproval: stage.RequiresApproval,
```
also nit: `RequiresApproval: env.Prod`

efekarakus commented:
This will be a breaking-change for customers that have a pipeline with environments that are initialized with `--prod`. When we unmarshal the manifest, the `requires_approval` field won't exist and the `copilot pipeline update` cmd will result in the removal of the manual approval.

I think this is fine though, we'll explicitly call it out in our release notes as this change gives the customers power to control the manual approvals :)

tam0ri committed:
Modify getEnvFromCache function.

Rename getEnvFromCache -> getEnvConfig.
Add break statement.

tam0ri committed:
Rename test function name

tam0ri committed:
Adress feedback

Rename CreatePipeline -> NewPipelineManifest.
Instead of passing []*config.Environment as input to NewPipelineManifest, passing []manifest.PipelineStage.

tam0ri committed:
Remove code added intentionally in previous commit.

kohidave commented on templates/cicd/pipeline.yml:
```diff
@@ -21,6 +21,8 @@ source:
     - # The name of the environment to deploy to.
       name: {{.Name}}
+      # Optional: flag for manual approval action before deployment.
+      requires_approval: {{.RequiresApproval}}
```
What do you think of having this commented out, by default (like the test commands)? It will be false by default. and folks can uncomment it if they want it!

tam0ri replied:
Sounds nice! I addressed.

tam0ri committed:
Modify getEnvConfig function.

tam0ri committed:
Modify pipeline.yml.

tam0ri committed:
Modify TestNewPipelineManifest.

efekarakus reviewed (approved):
🚀 Looks great! Thank you

mergify[bot] committed:
Merge branch 'mainline' into feat/pipeline-approval
